locust/test/test_main.py

```diff
@@ -1872,36 +2258,64 @@ def t(self):
                     "--expect-workers",
                     "1",
                     "-t",
-                    "1s",
+                    "3s",
                 ],
-                stderr=STDOUT,
+                stderr=PIPE,
                 stdout=PIPE,
                 text=True,
+                bufsize=1,
             )
-            gevent.sleep(2)
+
             proc_worker = subprocess.Popen(
                 [
                     "locust",
                     "-f",
                     "-",
                     "--worker",
                 ],
-                stderr=STDOUT,
+                stderr=PIPE,
                 stdout=PIPE,
                 text=True,
+                bufsize=1,
             )
 
-            stdout = proc.communicate()[0]
-            stdout_worker = proc_worker.communicate()[0]
+            output = []
 
-            self.assertIn('All users spawned: {"User1": 1} (1 total users)', stdout)
-            self.assertIn("Shutting down (exit code 0)", stdout)
-            self.assertNotIn("Traceback", stdout)
-            self.assertNotIn("Traceback", stdout_worker)
+            poll_until(lambda: all_users_spawned(proc, output), timeout=20)
+
+            end_time = time.time() + 15
+            while time.time() < end_time:
+                read_nonblocking(proc, output)
+                if "Shutting down" in "\n".join(output):
+                    break
+                time.sleep(0.1)
+
+            stdout, stderr = proc.communicate(timeout=20)
+            stdout_worker, stderr_worker = proc_worker.communicate(timeout=5)
+
+            output.extend(stdout.splitlines())
+            output.extend(stderr.splitlines())
+            output.extend(stdout_worker.splitlines())
+            output.extend(stderr_worker.splitlines())
+
+            self.assertTrue(
+                any("Shutting down" in line for line in output),
+                f"'Shutting down' not found in output:\n{''.join(output)}",
+            )
+            self.assertNotIn("Traceback", "\n".join(output))
 
             self.assertEqual(0, proc.returncode)
             self.assertEqual(0, proc_worker.returncode)
 
+            for p in [proc, proc_worker]:
```

Comment:
same here, there's no way these processes still alive this point, no need to kill them.